src/ooxml/java/org/apache/poi/xssf/usermodel/helpers/XSSFColumnShifter.java

```diff
@@ -0,0 +1,170 @@
+                            cloneCellValue(newCell, oldCell);
+                        }
+                        else {
+                            oldCell.setCellType(CellType.STRING);
+                            oldCell.setCellValue("");
+                        }
+                    }
+                    else {
+                        oldCell = row.createCell(columnIndex);
+                        if(newCell != null){
+                            oldCell.setCellType(newCell.getCellType());
+                            cloneCellValue(newCell, oldCell);
+                        }
+                        else {
+                            oldCell.setCellType(CellType.STRING);
+                            oldCell.setCellValue("");
+                        }
+                    }
+                }
+            }
+        }
+    }
+    
+    public void shiftComments(XSSFVMLDrawing vml, int startColumnIndex, int endColumnIndex, final int n, CommentsTable sheetComments){
+        SortedMap<XSSFComment, Integer> commentsToShift = new TreeMap<XSSFComment, Integer>(new Comparator<XSSFComment>() {
+            @Override
+            public int compare(XSSFComment o1, XSSFComment o2) {
+                int column1 = o1.getColumn();
+                int column2 = o2.getColumn();
+                
+                if(column1 == column2) {
+                    // ordering is not important when column is equal, but don't return zero to still 
+                    // get multiple comments per column into the map
+                    return o1.hashCode() - o2.hashCode();
+                }
+
+                // when shifting down, sort higher column-values first
+                if(n > 0) {
+                    return column1 < column2 ? 1 : -1;
+                } else {
+                    // sort lower-column values first when shifting up
+                    return column1 > column2 ? 1 : -1;
+                }
+            }
+        });
+        
+        if(sheetComments != null){
+            CTCommentList lst = sheetComments.getCTComments().getCommentList();
+            for (CTComment comment : lst.getCommentArray()) {
+                String oldRef = comment.getRef();
+                CellReference ref = new CellReference(oldRef);
+                
+                int newColumnIndex = XSSFShiftingManager.shiftedItemIndex(startColumnIndex, endColumnIndex, n, ref.getCol());
+                
+                // is there a change necessary for the current row?
+                if(newColumnIndex != ref.getCol()) {
+                    XSSFComment xssfComment = new XSSFComment(sheetComments, comment,
+                            vml == null ? null : vml.findCommentShape(ref.getRow(), ref.getCol()));
+                    commentsToShift.put(xssfComment, newColumnIndex);
+                }
+            }
+            for(Map.Entry<XSSFComment, Integer> entry : commentsToShift.entrySet()) 
```

Comment:
Do you need to accumulate comments into a commentsToShift map and then process that map? Can you modify the comments as you iterate over them, saving you from having to iterate over the comments twice?